### DISCUSSION TITLE
node.js -> Mongoose C++ HTTP server. How do I call/send a JS module ?

### DISCUSSION THREAD

**Discussion by linuxguy123** [cesanta/mongoose#1611] [Ask questions, get answers] [answered]
I'm transitioning some node.js code I wrote to the Mongoose C/C++ web server.

I want to serve a static HTML page that calls a JavaScript module that is in a separate file. ie Page1.html has the following in it.

`<script type="module" src="module.js"> </script>`

My Mongoose server serves up Page1.html, no problem. But Firefox throws an error when it loads the JS module:

`Loading module from “http://localhost:3000/module.js” was blocked because of a 
disallowed MIME type (“text/html”).`

I had the same problem in Node.js, but solved it by running node-static.

https://www.npmjs.com/package/node-static

`const static = require('node-static');
...
var file = new(static.Server)(__dirname);
...
const server = http.createServer(function (req,res){
    file.serve(req,res);
}).listen(3000);`

I have set mime_types:

`struct mg_http_serve_opts opts = {.mime_types ="html=text/html,js=text/javascript"};`

How do I achieve the same functionality in Mongoose ?  

Thanks

**Comment by scaprile** [comment]
The [device-dashboard](https://github.com/cesanta/mongoose/tree/master/examples/device-dashboard) example serves `<script type="module" src="main.js"></script>` with a static server and no extra options
```c
  } else if (ev == MG_EV_HTTP_MSG) {
    struct mg_http_message *hm = (struct mg_http_message *) ev_data;
    ...
    if (mg_http_match_uri(hm, "/api/hi")) {
      ...
    } else {
      struct mg_http_serve_opts opts = {0};
      opts.root_dir = "web_root";
      mg_http_serve_dir(c, ev_data, &opts);
    }
  }
```
Tutorials [here](https://mongoose.ws/tutorials/)

**Reply by linuxguy123** [reply]
OK, thanks for the example.

I saw that Firefox automatically does a second GET to retrieve module.js.   With my code it transfers 104 bytes and then stops.  It also does a GET for favicon.ico but transfers it without any issue.

Why can't module.js be transferred with my code if you allow its mime type ?

**Reply by scaprile** [reply]
Do the examples work for you ?
html and js file extensions have its MIME type already defined:
```c
    MG_C_STR("html"), MG_C_STR("text/html; charset=utf-8"),
    MG_C_STR("htm"), MG_C_STR("text/html; charset=utf-8"),
    MG_C_STR("css"), MG_C_STR("text/css; charset=utf-8"),
    MG_C_STR("js"), MG_C_STR("text/javascript; charset=utf-8"),
```
you shouldn't have to override them (the example does not, and you are not providing charset info in your code)
Who stops ? Please provide a minimal example and a Wireshark capture

**Reply by linuxguy123** [reply]
Thanks for the reply. 

I have not tested the code you supplied.  No doubt it would work, but I am unsure why my code doesn't work.  I am hesitant to allow serve-dir because I feel that is a security risk.

#tshark -i lo -d tcp.port==3000,http -Y http
   46 252.056036857    127.0.0.1 → 127.0.0.1    HTTP 533 GET /page1.html HTTP/1.1 
   48 252.056136207    127.0.0.1 → 127.0.0.1    HTTP 529 HTTP/1.1 200 OK  (text/html)
   50 252.424448067    127.0.0.1 → 127.0.0.1    HTTP 451 GET /module.js HTTP/1.1 
   52 252.424530388    127.0.0.1 → 127.0.0.1    HTTP 529 HTTP/1.1 200 OK  (text/html)

Note that the seconnd reply to the browser is tagged text/html, not text/javascript as it needs to be.

If you have a better wireshark command to gather data, I'll gladly run it.

**Reply by linuxguy123** [reply]
re: html and js file extensions have its MIME type already defined:

    `MG_C_STR("html"), MG_C_STR("text/html; charset=utf-8"),
    MG_C_STR("htm"), MG_C_STR("text/html; charset=utf-8"),
    MG_C_STR("css"), MG_C_STR("text/css; charset=utf-8"),
    MG_C_STR("js"), MG_C_STR("text/javascript; charset=utf-8"),`

I just tested with `struct mg_http_serve_opts opts = {0};` and got the same result. 

My file is UTF-8.   Do I need to have something special in the JS file for serve_file to recognize it as JS ?

I'll create a minimal app and supply source files in a bit.

**Reply by linuxguy123** [reply]
For what it is worth, I think it would be helpful if more details on mg_http_serve_opts were given in the API.   For example, I did not know that several types were set by default.

I know that generating and updating doc is a never ending (and thankless) task and I appreciate all the work that has been done to generate what there is.

**Reply by linuxguy123** [reply]
Minimal case

[server.zip](https://github.com/cesanta/mongoose/files/9023801/server.zip)

**Comment by scaprile** [comment, marked as answer]
Your code serves the very same file for every request, when you ask for module.js the server will send page1.html because that is what you are telling it to do.
Please start from a [tutorial](https://mongoose.ws/tutorials/http-server/) and work your way up.

**Reply by linuxguy123** [reply]
OK.

I was thinking that serve_file served the first file and all the subsequent files that the first file needed, ie module.js, automatically like node-static did in node.js.   I didn't realize that my code would need to reply to the request for the subsequent files as well.

If I use serve_dir, am I creating a security risk because an outsider could/would have access to any file in the serving directory ?

**Reply by scaprile** [reply]
put all your web files in _web_root_ as the example in the tutorial does and serve that dir.
if you require authorization, use it. The link in the first answer does that.
if you want to only serve specific files, match the URI and serve that file as the example in the documentation shows

**Reply by linuxguy123** [reply]
Thanks for your replies.

How do I get the URI for the GET from inside my event handler ?  For example, my client issues:
`101 1059.229776091    127.0.0.1 → 127.0.0.1    HTTP 451 GET /module.js HTTP/1.1`

How do I get 'module.js' from inside my event handler ?  
How do I get the URI from event_data ?  Where is event_data defined for HTTP ?

Update
I found it !   
ev_data is defined in the Events table on this page:
https://mongoose.ws/documentation/#event-handler-function

For MG_EV_HTTP_MSG,  ev_data is a pointer to struct mg_http_message.

`
struct mg_http_message {
struct mg_str method, uri, query, proto;             // Request/response line
struct mg_http_header headers[MG_MAX_HTTP_HEADERS];  // Headers
struct mg_str body;                                  // Body
 struct mg_str message;                               // Request line + headers + body
`

**Reply by linuxguy123** [reply]
I got it working.
I pulled the URI out of the HTTP message and stripped off the leading '/'.  That gives me the file name requested by the client.
Then I send that file with the serve_file function.  
I did not use the serve_dir function as the provided example used.
